synapse/handlers/pagination.py

```diff
@@ -72,13 +72,129 @@ def __init__(self, hs):
         self.auth = hs.get_auth()
         self.store = hs.get_datastore()
         self.clock = hs.get_clock()
+        self.config = hs.config
 
         self.pagination_lock = ReadWriteLock()
         self._purges_in_progress_by_room = set()
         # map from purge id to PurgeStatus
         self._purges_by_id = {}
         self._event_serializer = hs.get_event_client_serializer()
 
+        if self.config.retention_enabled:
+            # Run the purge jobs described in the configuration file.
+            for job in self.config.retention_purge_jobs:
+                self.clock.looping_call(
+                    self.purge_history_for_rooms_in_range,
+                    job["interval"],
+                    job["min_lifetime"], job["max_lifetime"],
+                )
+
+    @defer.inlineCallbacks
+    def is_event_outdated(self, event_ts, room_id):
+        """Checks whether the event with a given timestamp should be considered outdated
+        according to the room's or server's retention policy.
+
+        Args:
+            event_ts (int): Timestamp of the event in milliseconds.
+            room_id (str): Room in which the event was sent.
+
+        Returns:
+            bool: True if the event should be considered as outdated, False otherwise.
+        """
+        if self.config.retention_enabled:
+            try:
+                retention_policy = yield self.store.get_retention_policy_for_room(
+                    room_id
+                )
+
+                max_lifetime = retention_policy.get("max_lifetime")
+                if max_lifetime is None:
```

Comment:
if `retention_max_lifetime` is a cap rather than a fallback, this seems incorrect?

Reply:
The current requirement from DINSIC is to have those options acting both as a fallback and as a cap, e.g. if `max_lifetime == 1y` a user shouldn't be able to set a `max_lifetime > 1y` and a room without any policy should be considered as if `max_lifetime = 1y`. I believe this is described in https://github.com/matrix-org/matrix-dinsic/issues/337